From: Guido Gunther <agx@sigxcpu.org>
Date: Thu, 23 Dec 2021 23:00:00 +0100
Subject: [PATCH] Migrate to G_ADD_PRIVATE / get_instance_private

Backport of upstream post-1.2.0 commits (adapted to apply on 1.2.0):
  b7a8dd06 Use G_DEFINE_TYPE_WITH_PRIVATE instead of g_type_class_add_private
  bbe47ea8 Use nm_iodine_plugin_get_instance_private
  c2fe15d3 properties: Use iodine_editor_get_instance_private()

Fixes FTBFS with modern glib where G_TYPE_INSTANCE_GET_PRIVATE and
g_type_class_add_private are deprecated and -Werror is the default.

Upstream: https://gitlab.gnome.org/GNOME/network-manager-iodine
---
--- a/src/nm-iodine-service.c
+++ b/src/nm-iodine-service.c
@@ -50,15 +50,13 @@
 #define NM_IODINE_USER "nm-iodine"
 #define NM_IODINE_RUNDIR LOCALSTATEDIR "/run/" NM_IODINE_USER
 
-G_DEFINE_TYPE (NMIodinePlugin, nm_iodine_plugin, NM_TYPE_VPN_SERVICE_PLUGIN)
-
 typedef struct {
 	GPid pid;
 	NMVpnPluginFailure failure;
 	GVariantBuilder ip4config;
 } NMIodinePluginPrivate;
 
-#define NM_IODINE_PLUGIN_GET_PRIVATE(o) (G_TYPE_INSTANCE_GET_PRIVATE ((o), NM_TYPE_IODINE_PLUGIN, NMIodinePluginPrivate))
+G_DEFINE_TYPE_WITH_PRIVATE (NMIodinePlugin, nm_iodine_plugin, NM_TYPE_VPN_SERVICE_PLUGIN)
 
 static const char *iodine_binary_paths[] =
 {
@@ -247,7 +245,7 @@
 iodine_parse_stderr_line (NMVpnServicePlugin *plugin,
                           const char* line)
 {
-	NMIodinePluginPrivate *priv = NM_IODINE_PLUGIN_GET_PRIVATE (plugin);
+	NMIodinePluginPrivate *priv = nm_iodine_plugin_get_instance_private (NM_IODINE_PLUGIN (plugin));
 	gchar **split = NULL;
 	GVariant *val;
 	gint len;
@@ -338,7 +336,7 @@
 	GError *err = NULL;
 	gchar *line;
 	gint ret, l;
-	NMIodinePluginPrivate *priv = NM_IODINE_PLUGIN_GET_PRIVATE (plugin);
+	NMIodinePluginPrivate *priv = nm_iodine_plugin_get_instance_private (NM_IODINE_PLUGIN (plugin));
 
 	status = g_io_channel_read_line (source, &line, NULL, NULL, &err);
 	if (status != G_IO_STATUS_NORMAL) {
@@ -365,7 +363,7 @@
 iodine_watch_cb (GPid pid, gint status, gpointer user_data)
 {
 	NMIodinePlugin *plugin = NM_IODINE_PLUGIN (user_data);
-	NMIodinePluginPrivate *priv = NM_IODINE_PLUGIN_GET_PRIVATE (plugin);
+	NMIodinePluginPrivate *priv = nm_iodine_plugin_get_instance_private (NM_IODINE_PLUGIN (plugin));
 	guint error = 0;
 
 	if (WIFEXITED (status)) {
@@ -433,6 +431,7 @@
 	GIOChannel *stderr_channel;
 	gint stdin_fd, stderr_fd;
 	const char *props_topdomain, *props_fragsize, *props_nameserver;
+	NMIodinePluginPrivate *priv = nm_iodine_plugin_get_instance_private (NM_IODINE_PLUGIN (plugin));
 
 	/* Find iodine */
 	iodine_binary = iodine_binary_paths;
@@ -505,7 +504,7 @@
 	               iodine_stderr_cb,
 	               plugin);
 
-	NM_IODINE_PLUGIN_GET_PRIVATE (plugin)->pid = pid;
+	priv->pid = pid;
 	iodine_watch = g_child_watch_source_new (pid);
 	g_source_set_callback (iodine_watch,
 	                       (GSourceFunc) iodine_watch_cb,
@@ -585,7 +584,7 @@
 static gboolean
 real_disconnect (NMVpnServicePlugin *plugin, GError **err)
 {
-	NMIodinePluginPrivate *priv = NM_IODINE_PLUGIN_GET_PRIVATE (plugin);
+	NMIodinePluginPrivate *priv = nm_iodine_plugin_get_instance_private (NM_IODINE_PLUGIN (plugin));
 
 	if (priv->pid) {
 		if (kill (priv->pid, SIGTERM) == 0)
@@ -603,7 +602,7 @@
 static void
 nm_iodine_plugin_init (NMIodinePlugin *plugin)
 {
-	NMIodinePluginPrivate *priv = NM_IODINE_PLUGIN_GET_PRIVATE (plugin);
+	NMIodinePluginPrivate *priv = nm_iodine_plugin_get_instance_private (NM_IODINE_PLUGIN (plugin));
 
 	g_variant_builder_init (&priv->ip4config, G_VARIANT_TYPE_VARDICT);
 	priv->failure = -1;
@@ -615,8 +614,6 @@
 	GObjectClass *object_class = G_OBJECT_CLASS (iodine_class);
 	NMVpnServicePluginClass *parent_class = NM_VPN_SERVICE_PLUGIN_CLASS (iodine_class);
 
-	g_type_class_add_private (object_class, sizeof (NMIodinePluginPrivate));
-
 	/* virtual methods */
 	parent_class->connect    = real_connect;
 	parent_class->need_secrets = real_need_secrets;

--- a/properties/nm-iodine.c
+++ b/properties/nm-iodine.c
@@ -84,12 +84,6 @@
 
 static void iodine_editor_interface_init (NMVpnEditorInterface *iface_class);
 
-G_DEFINE_TYPE_EXTENDED (IodineEditor, iodine_editor,G_TYPE_OBJECT, 0,
-                        G_IMPLEMENT_INTERFACE (NM_TYPE_VPN_EDITOR,
-                                               iodine_editor_interface_init))
-
-#define IODINE_EDITOR_GET_PRIVATE(o) (G_TYPE_INSTANCE_GET_PRIVATE ((o), IODINE_TYPE_EDITOR, IodineEditorPrivate))
-
 typedef struct {
 	GtkBuilder *builder;
 	GtkWidget *widget;
@@ -97,6 +91,11 @@
 	gboolean window_added;
 } IodineEditorPrivate;
 
+G_DEFINE_TYPE_WITH_CODE (IodineEditor, iodine_editor, G_TYPE_OBJECT,
+                         G_ADD_PRIVATE (IodineEditor)
+                         G_IMPLEMENT_INTERFACE (NM_TYPE_VPN_EDITOR,
+                                                iodine_editor_interface_init))
+
 typedef enum {
 	NM_IODINE_IMPORT_EXPORT_ERROR_UNKNOWN = 0,
 	NM_IODINE_IMPORT_EXPORT_ERROR_NOT_IODINE,
@@ -252,7 +251,7 @@
 check_validity (IodineEditor *self, GError **error)
 {
 	IodineEditorPrivate *priv = \
-		IODINE_EDITOR_GET_PRIVATE (self);
+		iodine_editor_get_instance_private (self);
 	GtkWidget *widget;
 	const char *str;
 
@@ -282,7 +281,7 @@
                        const char *secret_name)
 {
 	IodineEditorPrivate *priv = \
-		IODINE_EDITOR_GET_PRIVATE (self);
+		iodine_editor_get_instance_private (self);
 
 	NMSettingSecretFlags secret_flags = NM_SETTING_SECRET_FLAG_NONE;
 	GtkWidget *widget;
@@ -317,7 +316,7 @@
 static void
 show_toggled_cb (GtkCheckButton *button, IodineEditor *self)
 {
-	IodineEditorPrivate *priv = IODINE_EDITOR_GET_PRIVATE (self);
+	IodineEditorPrivate *priv = iodine_editor_get_instance_private (self);
 	GtkWidget *widget;
 	gboolean visible;
 
@@ -332,7 +331,7 @@
 pw_type_combo_changed_cb (GtkWidget *combo, gpointer user_data)
 {
 	IodineEditor *self = IODINE_EDITOR (user_data);
-	IodineEditorPrivate *priv = IODINE_EDITOR_GET_PRIVATE (self);
+	IodineEditorPrivate *priv = iodine_editor_get_instance_private (self);
 	GtkWidget *entry;
 
 	entry = GTK_WIDGET (gtk_builder_get_object (priv->builder, "password_entry"));
@@ -362,7 +361,7 @@
                    const char *secret_key,
                    const char *entry_name)
 {
-	IodineEditorPrivate *priv = IODINE_EDITOR_GET_PRIVATE (self);
+	IodineEditorPrivate *priv = iodine_editor_get_instance_private (self);
 	int active = -1;
 	GtkWidget *widget;
 	GtkListStore *store;
@@ -415,7 +414,7 @@
                 NMConnection *connection,
                 GError **error)
 {
-	IodineEditorPrivate *priv = IODINE_EDITOR_GET_PRIVATE (self);
+	IodineEditorPrivate *priv = iodine_editor_get_instance_private (self);
 	NMSettingVpn *s_vpn;
 	GtkWidget *widget;
 	const char *value;
@@ -484,7 +483,7 @@
 get_widget (NMVpnEditor *iface)
 {
 	IodineEditor *self = IODINE_EDITOR (iface);
-	IodineEditorPrivate *priv = IODINE_EDITOR_GET_PRIVATE (self);
+	IodineEditorPrivate *priv = iodine_editor_get_instance_private (self);
 
 	return G_OBJECT (priv->widget);
 }
@@ -533,7 +532,7 @@
                    GError **error)
 {
 	IodineEditor *self = IODINE_EDITOR (iface);
-	IodineEditorPrivate *priv = IODINE_EDITOR_GET_PRIVATE (self);
+	IodineEditorPrivate *priv = iodine_editor_get_instance_private (self);
 	NMSettingVpn *s_vpn;
 	GtkWidget *widget;
 	char *str;
@@ -593,7 +592,7 @@
 		return NULL;
 	}
 
-	priv = IODINE_EDITOR_GET_PRIVATE (object);
+	priv = iodine_editor_get_instance_private (IODINE_EDITOR (object));
 	ui_file = g_strdup_printf ("%s/%s", UIDIR, "nm-iodine-dialog.ui");
 	priv->builder = gtk_builder_new ();
 
@@ -632,7 +631,7 @@
 {
 	IodineEditor *plugin = IODINE_EDITOR (object);
 	IodineEditorPrivate *priv = \
-		IODINE_EDITOR_GET_PRIVATE (plugin);
+		iodine_editor_get_instance_private (IODINE_EDITOR (plugin));
 
 	if (priv->group)
 		g_object_unref (priv->group);
@@ -651,8 +650,6 @@
 {
 	GObjectClass *object_class = G_OBJECT_CLASS (req_class);
 
-	g_type_class_add_private (req_class, sizeof (IodineEditorPrivate));
-
 	object_class->dispose = dispose;
 }
 
